### PR TITLE
Exception for non supported header

### PR DESCRIPTION
**Pull request by chunyu3**
# Description
<i>Add your description here!</i>

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs [hunk 1]
```diff
@@ -168,6 +368,97 @@ internal HttpMessage CreateCollapseGetRequest(RequestConditions requestCondition
...
+            var message = _pipeline.CreateMessage(context);
+            var request = message.Request;
+            request.Method = RequestMethod.Get;
+            var uri = new RawRequestUriBuilder();
+            uri.Reset(_endpoint);
+            uri.AppendPath("/RequestConditionCollapse/missIfMatch", false);
+            request.Uri = uri;
+            if (requestConditions != null)
+            {
+                if (requestConditions.IfMatch != null) throw new NotSupportedException($"IfMatch head is not supported here");
```
Comment (AlexanderSher) [inline test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs:400]: Supported values are the feature of parameter, not header. Hence this check should happen in `MissIfMatchGet`/`MissIfMatchGetAsync` methods, and exception type will be `ArgumentException`. Ideally, I'd like to have a method in `Azure.Core.Argument`, similar to `Argument.AssertNotNullOrEmpty`, something like:
```cs
public static void AssertHasOnlySupportedHeaders(RequestConditions? requestConditions, string paramName, params string[] supportedHeaders)
{
    if (requestConditions != null)
    {
        AssertHasNoUnsupportedHeader(requestConditions.IfMatch, paramName, supportedHeaders, HttpHeader.Names.IfMatch);
        AssertHasNoUnsupportedHeader(requestConditions.IfNoneMatch, paramName, supportedHeaders, HttpHeader.Names.IfNoneMatch);
        AssertHasNoUnsupportedHeader(requestConditions.IfModifiedSince, paramName, supportedHeaders, HttpHeader.Names.IfModifiedSince);
        AssertHasNoUnsupportedHeader(requestConditions.IfUnmodifiedSince, paramName, supportedHeaders, HttpHeader.Names.IfUnmodifiedSince);
    }
}

private static void AssertHasNoUnsupportedHeader<T>(T? value, string paramName, string[] supportedHeaders, string unsupportedHeader) where T : struct
{
    if (value.HasValue && !supportedHeaders.Contains(unsupportedHeader))
    {
        throw new ArgumentException($"{unsupportedHeader} header is not supported. Supported headers are: {string.Join(", ", supportedHeaders)}.", paramName);
    }
}
```
```cs
public virtual async Task<Response> MissIfMatchGetAsync(RequestConditions requestConditions = null, RequestContext context = null) 
{
    Argument.AssertHasOnlySupportedHeaders(requestConditions, nameof(requestConditions), "If-None-Match", "If-Modified-Since");
```

This approach would have several benefits:
- testability
- less generated code
- no need to regenerate old code if another property is added to `RequestConditions`

Comment (AlexanderSher) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs:400]: The question however is how to add validation information to the `AutoRest.CSharp.Output.Models.Shared.Parameter`, since we don't have any support for custom validation. Let me think about it.

File: test/TestProjects/CollapseRequestCondition-LowLevel/CollapseRequestCondition-LowLevel.json [hunk 2]
```diff
@@ -343,18 +450,46 @@
               "name": "modified-access-conditions"
             },
             "description": "Specify an ETag value to operate only on blobs with a matching value."
-          }  
+          },
+	        {
```
Comment (AlexanderSher) [inline test/TestProjects/CollapseRequestCondition-LowLevel/CollapseRequestCondition-LowLevel.json:454]: Formatting

Comment (chunyu3) [reply test/TestProjects/CollapseRequestCondition-LowLevel/CollapseRequestCondition-LowLevel.json:454]: resolved

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs (hunk 1 above)
Comment (chunyu3) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs:400]: Yes, currently we have not put parameter validation in the rest Method, and currently we have put some validation in Create****Request, such as check if a header is NULL when try to add the Head to HTTP Request. We can put the validation either location, but we need to consistent.

File: src/assets/Azure.Core.Shared/Argument.cs [hunk 3]
```diff
@@ -212,5 +212,24 @@ public static string CheckNotNullOrEmpty(string value, string name)
             AssertNotNullOrEmpty(value, name);
             return value;
         }
+
+        public static void AssertHasOnlySupportedHeaders(RequestConditions? requestConditions, string paramName, string[] supportedHeaders)
```
Comment (annelo-msft) [inline src/assets/Azure.Core.Shared/Argument.cs:216]: I believe files under Azure.Core.Shared are copied in from the azure-sdk-for-net repo, so this change may need to be made there.

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/NonCollapseClient.cs [hunk 4]
```diff
@@ -88,17 +88,18 @@ public virtual Response CollapsePut(RequestContent content, ETag? ifMatch = null
             }
         }
 
-        /// <param name="requestConditions"> The content to send as the request conditions of the request. </param>
+        /// <param name="content"> The content to send as the body of the request. </param>
+        /// <param name="ifNoneMatch"> Specify an ETag value to operate only on blobs without a matching value. </param>
         /// <param name="context"> The request context, which can override default behaviors on the request on a per-call basis. </param>
 #pragma warning disable AZC0002
-        public virtual async Task<Response> CollapseGetAsync(RequestConditions requestConditions = null, RequestContext context = null)
+        public virtual async Task<Response> IfNonMatchPutAsync(RequestContent content, ETag? ifNoneMatch = null, RequestContext context = null)
```
Comment (annelo-msft) [inline test/TestProjects/CollapseRequestCondition-LowLevel/Generated/NonCollapseClient.cs:95]: nit: spelling - I called this out in the .json file as well.

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs [hunk 5]
```diff
@@ -164,6 +364,97 @@ internal HttpMessage CreateCollapseGetRequest(RequestConditions requestCondition
...
+            var message = _pipeline.CreateMessage(context);
+            var request = message.Request;
+            request.Method = RequestMethod.Get;
+            var uri = new RawRequestUriBuilder();
+            uri.Reset(_endpoint);
+            uri.AppendPath("/RequestConditionCollapse/missIfModifiedSince", false);
+            request.Uri = uri;
+            if (requestConditions != null)
+            {
+                if (requestConditions.IfModifiedSince != null) throw new NotSupportedException($"IfModifiedSince head is not supported here");
```
Comment (annelo-msft) [inline test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs:414]: Parameter validations should happen inside the client call, before other actions are executed.  (For example, we don't want to start a Diagnostic Scope, if we know the method is going to fail with an ArgumentException.)

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs (hunk 1 above)
Comment (annelo-msft) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs:396]: I'd like to simplify what goes into Argument.cs so that the generated code looks more like this:

```
public virtual Response DoThing(RequestConditions conditions, RequestContext context = default)
{
  Argument.AssertNull(requestConditions.IfMatch, nameof(requestConditions, "Service does not support the If-Match header for this operation.");
  // the rest of the service call
}

**Review by annelo-msft — COMMENTED** [review]
I left suggested changes.  I'd also like to get @ShivangiReja's review on the code under /src after my comments are addressed.

File: src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs [hunk 6]
```diff
@@ -15,6 +15,7 @@
 using Azure.Core;
 using AutoRest.CSharp.Utilities;
 using Azure;
+using AutoRest.CSharp.Common.Output.Models;
```
Comment (ShivangiReja) [inline src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs:18]: Nit: Why is this needed? nothing changed in this file.

File: test/TestProjects/CollapseRequestCondition-LowLevel/CollapseRequestCondition-LowLevel.json [hunk 7]
```diff
@@ -402,4 +565,4 @@
       }
     }
   }
-}
+}
```
Comment (ShivangiReja) [inline test/TestProjects/CollapseRequestCondition-LowLevel/CollapseRequestCondition-LowLevel.json:568]: Nit: Add new line at the end of the file

File: test/AutoRest.Shared.Tests/AutoRest.Shared.Tests.csproj [hunk 8]
```diff
@@ -1,14 +1,13 @@
...
-    <TargetFramework>netcoreapp3.1</TargetFramework>
+    <TargetFrameworks>netcoreapp3.1;net6.0</TargetFrameworks>
     <IsPackable>false</IsPackable>
     <Nullable>annotations</Nullable>
-    <NoWarn>SA1649;SA1633</NoWarn>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Azure.Core" Version="1.22.0" />
+    <PackageReference Include="Azure.Core" Version="1.22.0-alpha.20211207.1" />
```
Comment (ShivangiReja) [inline test/AutoRest.Shared.Tests/AutoRest.Shared.Tests.csproj:10]: Why are we using different version of Azure.Core only for tests?

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/MatchConditionCollapseClient.cs [hunk 9]
```diff
@@ -267,23 +267,6 @@ internal HttpMessage CreateCollapseGetRequest(MatchConditions matchConditions, R
...
-            uri.AppendPath("/MatchConditionCollapse/multi", false);
-            request.Uri = uri;
-            if (requestConditions != null)
-            {
-                request.Headers.Add(requestConditions, "R");
-            }
-            message.ResponseClassifier = ResponseClassifier200.Instance;
-            return message;
-        }
-
```
Comment (ShivangiReja) [inline test/TestProjects/CollapseRequestCondition-LowLevel/Generated/MatchConditionCollapseClient.cs:286]: Why is this method got removed? I think you will get compile time error because we are calling this method from  [MulticollapseGetAsync()](https://github.com/Azure/autorest.csharp/blob/cf0fce46867d697a0e6fa91ea9525b0c370bfbad/test/TestProjects/CollapseRequestCondition-LowLevel/Generated/MatchConditionCollapseClient.cs#L183) and [MulticollapseGet()](https://github.com/Azure/autorest.csharp/blob/cf0fce46867d697a0e6fa91ea9525b0c370bfbad/test/TestProjects/CollapseRequestCondition-LowLevel/Generated/MatchConditionCollapseClient.cs#L203).

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs [hunk 10]
```diff
@@ -886,7 +888,7 @@ private void AddRequestParameter(string name, RequestParameter requestParameter,
                 }
             }
         }
-
+        /*
         [Flags]
         private enum RequestConditionHeaders
```
Comment (ShivangiReja) [inline src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs:893]: We are adding `RequestConditionHeaders.cs` for enums, can we remove this commented code?

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs [hunk 11]
```diff
@@ -128,6 +128,206 @@ public virtual Response CollapseGet(RequestConditions requestConditions = null,
             }
         }
 
+        /// <param name="requestConditions"> The content to send as the request conditions of the request. </param>
+        /// <param name="context"> The request context, which can override default behaviors on the request on a per-call basis. </param>
+#pragma warning disable AZC0002
+        public virtual async Task<Response> MissIfNonMatchGetAsync(RequestConditions requestConditions = null, RequestContext context = null)
+#pragma warning restore AZC0002
+        {
+            using var scope = _clientDiagnostics.CreateScope("RequestConditionCollapseClient.MissIfNonMatchGet");
```
Comment (ShivangiReja) [inline test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs:137]: Shouldn't it generate `_clientDiagnostics` -> `ClientDiagnostics` ?

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs (hunk 1 above)
Comment (chunyu3) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs]: hi @annelo-msft 
When there are more than one request condition need to check, the Argument.AssertNull will be call more than once.
e..g 

public virtual Response DoThing(RequestConditions conditions, RequestContext context = default)
{
    Argument.AssertNull(requestConditions.IfMatch, nameof(requestConditions), "Service does not support the If-Match header for this operation.");
  Argument.AssertNull(requestConditions.IfNoneMatch, nameof(requestConditions), "Service does not support the If-None-Match header for this operation.");
    // the rest of the service call
}

The worst case, the Argument.AssertNull will call three times.
Is it acceptable?

File: src/assets/Azure.Core.Shared/Argument.cs (hunk 3 above)
Comment (chunyu3) [reply src/assets/Azure.Core.Shared/Argument.cs:216]: Yes, we need to update in azure-sdk-for-net repo, after we merge those changes into azure-sdk-for-net repo, we need first file an PR in autorest.csharp to update the Argument.cs which is copied from azure-sdk-for-net. Then file another PR to include our business logic. We can merge those two PR as one.

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs (hunk 1 above)
Comment (annelo-msft) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs]: I think this is a very edge case where we have these asserts in the first place.  I think simplicity of Azure.Core APIs and readability of the code should be the driving factor, and we can revisit the perf implications if we find we have a problem in the future.  This would give us a strong motivating reason to write more complex APIs into the Assert class.

We will not add properties to RequestConditions often -- this type represents [HTTP conditional requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests), and has [four of the five headers](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#conditional_headers) represented.  We haven't yet found a case in Azure that needed the fifth - we would only add it if we did, and then there would be no further headers to add.

And @AlexanderSher, I don't understand the argument about having to regenerate code -- don't we always regenerate code when we make changes to the generator?

Comment (chunyu3) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs]: When we add a new field into RequestConditions type, then we need to also check if the new field is set or not. In that way, we need to regenerated the existing SDK client libraries to reflect it. 
But if we use AssertHasOnlySupportedHeaders, we just need to update the method in Azure.Core, the existing SDK client libraries does not updated.

Comment (chunyu3) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs]: Yes, RequestConditions type is for HTTP conditional requests, so it is rare or maybe not to be updated (such as add or remove an field), so we may not need to consider this scenario current. Right? @annelo-msft

Comment (annelo-msft) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs]: We'll never remove a field, because that would be a breaking change, and there is only one field left to add, so we would only ever regenerate once.  So, yes, I wouldn't worry about the possible addition - we regenerate the library code fairly often anyway.

Comment (chunyu3) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs]: I will implement it this way.

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs (hunk 10 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs:905]: removed

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/MatchConditionCollapseClient.cs (hunk 9 above)
Comment (chunyu3) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/MatchConditionCollapseClient.cs:273]: resolved

Comment (chunyu3) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/MatchConditionCollapseClient.cs:273]: resolved

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs (hunk 11 above)
Comment (chunyu3) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs]: resolved

File: src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs (hunk 10 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Output/Models/RestClientBuilder.cs]: removed

File: src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs (hunk 6 above)
Comment (chunyu3) [reply src/AutoRest.CSharp/Common/Generation/Writers/RequestWriterHelpers.cs]: removed

File: test/TestProjects/CollapseRequestCondition-LowLevel/CollapseRequestCondition-LowLevel.json (hunk 7 above)
Comment (chunyu3) [reply test/TestProjects/CollapseRequestCondition-LowLevel/CollapseRequestCondition-LowLevel.json]: resolved

File: test/AutoRest.Shared.Tests/AutoRest.Shared.Tests.csproj (hunk 8 above)
Comment (chunyu3) [reply test/AutoRest.Shared.Tests/AutoRest.Shared.Tests.csproj]: resolved

File: test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs (hunk 1 above)
Comment (annelo-msft) [reply test/TestProjects/CollapseRequestCondition-LowLevel/Generated/RequestConditionCollapseClient.cs]: Thanks, @chunyu3!

**Review by annelo-msft — APPROVED** [review]
The output in RequestConditionCollapseClient looks correct to me.  @AlexanderSher or @ShivangiReja, does the generator source look good to you?
